ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/PerasCertDB/Impl.hs

```diff
@@ -0,0 +1,297 @@
+  | PerasCertDbClosed
+  deriving stock Generic
+  deriving anyclass NoThunks
+
+data PerasCertDbEnv m blk = PerasCertDbEnv
+  { pcdbTracer :: !(Tracer m (TraceEvent blk))
+  , pcdbVolatileState :: !(StrictTVar m (WithFingerprint (PerasVolatileCertState blk)))
+  -- ^ The 'RoundNo's of all certificates currently in the db.
+  }
+  deriving NoThunks via OnlyCheckWhnfNamed "PerasCertDbEnv" (PerasCertDbEnv m blk)
+
+getEnv ::
+  (IOLike m, HasCallStack) =>
+  PerasCertDbHandle m blk ->
+  (PerasCertDbEnv m blk -> m r) ->
+  m r
+getEnv (PerasCertDbHandle varState) f =
+  readTVarIO varState >>= \case
+    PerasCertDbOpen env -> f env
+    PerasCertDbClosed -> throwIO $ ClosedDBError prettyCallStack
+
+getEnv1 ::
+  (IOLike m, HasCallStack) =>
+  PerasCertDbHandle m blk ->
+  (PerasCertDbEnv m blk -> a -> m r) ->
+  a ->
+  m r
+getEnv1 h f a = getEnv h (\env -> f env a)
+
+getEnvSTM ::
+  (IOLike m, HasCallStack) =>
+  PerasCertDbHandle m blk ->
+  (PerasCertDbEnv m blk -> STM m r) ->
+  STM m r
+getEnvSTM (PerasCertDbHandle varState) f =
+  readTVar varState >>= \case
+    PerasCertDbOpen env -> f env
+    PerasCertDbClosed -> throwIO $ ClosedDBError prettyCallStack
+
+{-------------------------------------------------------------------------------
+  API implementation
+-------------------------------------------------------------------------------}
+
+implCloseDB :: IOLike m => PerasCertDbHandle m blk -> m ()
+implCloseDB (PerasCertDbHandle varState) =
+  atomically (swapTVar varState PerasCertDbClosed) >>= \case
+    PerasCertDbOpen PerasCertDbEnv{pcdbTracer} -> do
+      traceWith pcdbTracer ClosedPerasCertDB
+    -- DB was already closed.
+    PerasCertDbClosed -> pure ()
+
+-- TODO: we will need to update this method with non-trivial validation logic
+-- see https://github.com/tweag/cardano-peras/issues/120
+implAddCert ::
+  ( IOLike m
+  , StandardHash blk
+  ) =>
+  PerasCertDbEnv m blk ->
+  ValidatedPerasCert blk ->
+  m AddPerasCertResult
+implAddCert env cert = do
+  traceWith pcdbTracer $ AddingPerasCert roundNo boostedPt
+  res <- atomically $ do
+    WithFingerprint
+      PerasVolatileCertState
+        { pvcsCerts
+        , pvcsWeightByPoint
+        , pvcsCertsByTicket
+        , pvcsLastTicketNo
+        }
+      fp <-
+      readTVar pcdbVolatileState
+    if Map.member roundNo pvcsCerts
+      then pure PerasCertAlreadyInDB
+      else do
+        let pvcsLastTicketNo' = succ pvcsLastTicketNo
+        writeTVar pcdbVolatileState $
+          WithFingerprint
+            PerasVolatileCertState
+              { pvcsCerts =
+                  Map.insert roundNo cert pvcsCerts
+              , -- Note that the same block might be boosted by multiple points.
+                pvcsWeightByPoint =
+                  addToPerasWeightSnapshot boostedPt (getPerasCertBoost cert) pvcsWeightByPoint
+              , pvcsCertsByTicket =
+                  Map.insert pvcsLastTicketNo' cert pvcsCertsByTicket
```

Comment:
Thinking generally, `certsByTicket` sounds like a specific "index" over a certs "table" which has a `roundNo` as a primary key (if I read this correctly).

Sometimes I opt to maintain indices such that they point to object 'identifiers', rather than objects themselves `Map TicketNo RoundNo` so to speak. You probably see that there's 2 lookups that need to happen here to go from ticketNo to a cert.

Sometimes that makes sense, for example 1-to-may indices (which this isn't), if there are updates (which there aren't), and if objects are copied (which here they aren't...me thinks).